File: app/src/main/java/net/mbonnin/arcanetracker/Lce.java
package net.mbonnin.arcanetracker;

public class Lce<T> {
    private final Throwable mError;
    T mData;

    private Lce(T data, Throwable error) {
        this.mData = data;
        this.mError = error;
    }
    public static <T> Lce<T> data(T data) {
        return new Lce<T>(data, null);
    }
    public static <T> Lce<T> error(Throwable error) {
        return new Lce<T>(null, error);
    }
    public static <T> Lce<T> loading() {
        return new Lce<T>(null, null);
    }
    public boolean isLoading() {
        return mData == null && mError == null;
    }
    public Throwable getError() {
        return mError;
    }
    public T getData() {
        return mData;
    }
}

File: app/src/main/java/net/mbonnin/arcanetracker/hsreplay/model/Lce.kt
package net.mbonnin.arcanetracker.hsreplay.model

data class Lce<T> private constructor(val data: T?, val error: Throwable?) {
    val isLoading = data == null && error == null

    companion object {
        fun <T> data(data: T): Lce<T> {
            return Lce(data, null)
        }

        fun <T> error(error: Throwable): Lce<T> {
            return Lce<T>(null, error)
        }

        fun <T> loading(): Lce<T> {
            return Lce<T>(null, null)
        }
    }
}